src/hotspot/share/opto/memnode.cpp

```diff
@@ -3511,7 +3515,43 @@ Node* StoreNode::Identity(PhaseGVN* phase) {
       mem->in(MemNode::Address)->eqv_uncast(adr) &&
       mem->in(MemNode::ValueIn)->eqv_uncast(val) &&
       mem->Opcode() == Opcode()) {
-    result = mem;
+    // Not a vector
+    if (!is_StoreVector()) {
+      result = mem;
+    } else {
+      // Ensure that vector types match
+      const TypeVect* vect_type = as_StoreVector()->vect_type();
+      const TypeVect* mem_vect_type = mem->as_StoreVector()->vect_type();
+      if (vect_type == mem_vect_type) {
+        // Handle StoreVector with offsets and masks
+        // Ensure offsets match
+        if (is_StoreVectorScatter()) {
+          const Node* offsets = as_StoreVectorScatter()->in(StoreVectorScatterNode::Offsets);
+          if (offsets->eqv_uncast(mem->as_StoreVectorScatter()->in(StoreVectorScatterNode::Offsets))) {
+            result = mem;
+          }
+        // Ensure masks match
+        } else if (is_StoreVectorMasked()) {
+          const Node* mask = as_StoreVectorMasked()->in(StoreVectorMaskedNode::Mask);
+          if (mask->eqv_uncast(mem->as_StoreVectorMasked()->in(StoreVectorMaskedNode::Mask))) {
+            result = mem;
+          }
+        // Ensure offsets and masks match
+        } else if (is_StoreVectorScatterMasked()) {
+          const StoreVectorScatterMaskedNode* stv = as_StoreVectorScatterMasked();
+          const Node* offsets = stv->in(StoreVectorScatterMaskedNode::Offsets);
+          const Node* mask = stv->in(StoreVectorScatterMaskedNode::Mask);
+          const StoreVectorScatterMaskedNode* svgm = mem->as_StoreVectorScatterMasked();
+          if (offsets->eqv_uncast(svgm->in(StoreVectorScatterMaskedNode::Offsets)) &&
+            mask->eqv_uncast(svgm->in(StoreVectorScatterMaskedNode::Mask))) {
+            result = mem;
+          }
+        // Regular store (no offsets or mask)
+        } else {
+          result = mem;
+        }
+      }
+    }
```

Comment:
I agree that it is quite convoluted probably also because I've put `if (!is_StoreVector())` (which is redundant) at the beginning to get the most common case out of the way but still...
At first I thought that multiple inheritance would be a good solution (masks and offsets could be inherited by the corresponding nodes) but the "HotSpot Coding Style" clearly says to avoid it...
So, I think in the end your second suggestion is the cleanest. Changing it...

Reply:
I've updated it. The condition unfortunately doesn't look as clean as the one above as we need to check for `nullptr` (either both or none and `eqv_uncast`). I've tried to make it as concise as possible (we could have made `mask` and `offsets` return a _unique_ node instead, so as to avoid the `nullptr`, but I had the impression it would just make everything less clear).

Reply:
I think this is good, certainly much better than what we had before! :blush: